### DISCUSSION TITLE
Applications deployed using docker do not support. env files

### DISCUSSION THREAD

**Discussion by xiaoxiji1234** [lowdefy/lowdefy#970] [Questions and Help]
Hi, My lowdefy application after docker deployment cannot call the request or jump to the keycloak login interface. Does it not support. env files after deployment? Is there any solution?

**Comment by SamTolmay** [comment]
Hi @xiaoxiji1234 

We recommend that you use Docker to configure the environment variables. If you are using `docker run` you can use the `-e` or `-env-file` options as shown here: https://docs.docker.com/engine/reference/commandline/run/#set-environment-variables--e---env---env-file

If you are using an orchestration framework like Kubernetes or Docker Swarm you can use those to set the env variables.

**Reply by xiaoxiji1234** [reply]
Ok, Thank you very much!

**Comment by 13316090972** [comment]
My lowdefy app is intergrated with keycloak as an open-id provider, and works ok when running as a process like `npx lowdefy@latest dev --port 8848 ` or `lowdefy dev --port 8848`.
But it encounters an error as bellow when running on docker.
![image](https://user-images.githubusercontent.com/56707831/145329828-05031b4d-aab8-4198-9c6c-1060bc7e7379.png)


docker run command:
```
docker run -d \
  -p 8848:3000 \
  --name diclap \
  --mount type=volume,source=lowdefy-build,destination=/home/node/lowdefy/build \
  --env-file ./.env \
  lowdefy/lowdefy:latest
docker cp ./.lowdefy/build diclap:/home/node/lowdefy
```
Environment variables in docker :
```
[diclap@ecs-c7c3-0730395]$ docker exec -it diclap /bin/sh
~/lowdefy $ env
NODE_VERSION=14.18.1
HOSTNAME=afcd37aa4365
YARN_VERSION=1.22.15
SHLVL=1
HOME=/home/node
TERM=xterm
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
LOWDEFY_SECRET_OPENID_DOMAIN="http://xxx.xxx.xxx.xxx:9080/auth/realms/diclap-integrate"
LOWDEFY_SECRET_AXIOS_BASE_URL="http://xxx.xxx.xxx.xxx:1880/diclap/v1/"
LOWDEFY_SECRET_JWT_SECRET="secret"
PWD=/home/node/lowdefy
LOWDEFY_SECRET_OPENID_CLIENT_ID="diclap-integrate"
LOWDEFY_SECRET_OPENID_CLIENT_SECRET="secret"
NODE_ENV=production
```

**Reply by SamTolmay** [reply]
Hi @13316090972 

That error means the OpenID Connect Client hasn't received a valid url for the OpenID domain. We often see this when the `https://` protocol part of the domain is missing, but it looks like you have that configured correctly., so I am not sure what the issue is. 

Maybe you could look at the client library issues, which you can find here: https://github.com/panva/node-openid-client/issues

Please also change the secrets you shared here, as they are publically visible 😀

**Reply by 13316090972** [reply]
Thanks for your advise,I'll change the secrets on production server.

**Reply by 13316090972** [reply]
The problem was resolved yesterday.
Firstly,the docker network mode should be host mode but not the default bridge mode.Secondly, I found that the docker with "config-in-image" or "config-in-volumn" didn't send openid request (it confused me) by observing tcpdump, so i built a new lowdefy image as bellow.
```
FROM centos:latest AS build
# install nodejs and lowdefy
RUN dnf module install -y nodejs:14 \
    && npm install -g lowdefy  --unsafe-perm=true --allow-root \
    && mkdir -p /opt/app
#set workdir to /opt/app
WORKDIR /opt/app

#set workdir as a persistency directory
VOLUME ["/opt/app/"]
# Run lowdefy on start
CMD ["lowdefy", "dev"]
```
As you can see, the new lowdefy image is for dev enviroment but not for prod enviroment .

**Comment by 13316090972** [comment]
docker run command:
![image](https://user-images.githubusercontent.com/56707831/145329991-de83389b-9abf-4580-b4d8-d2e62576f926.png)
Environment variables in docker :
![image](https://user-images.githubusercontent.com/56707831/145330040-f086e6f6-96a2-4838-b1fa-5ed46963941d.png)
